### PR TITLE
Update two-way irrigation scheme to fit tri-grid configuration

### PR DESCRIPTION
**Pull request by hydrotian**
The two-way irrigation scheme between land and river models is updated to fit the tri-grid 
configuration. The temporary water conservation solution developed for different land and river 
grid is removed. Two additional flags are added to the land model to make the irrigation scheme 
more flexible: extra_gw_irr (for extra groundwater pumping) and firrig_data (for using 
prepossessed parameter).  Also some modifications were made to couple MOSART-WM and 
MOSART-inundation.

[BFB]

### CODE REVIEW

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 1]
```diff
@@ -462,8 +462,9 @@ subroutine Rtmini(rtm_active,flood_active)
     end if
 
     if (wrmflag .and. inundflag) then
-       write(iulog,*) subname,' ERROR MOSART wrmflag and inundflag cannot both be true'
-       call shr_sys_abort( subname//' ERROR: wrmflag and inundflag both set' )
+       !write(iulog,*) subname,' ERROR MOSART wrmflag and inundflag cannot both be true'
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: delete commented out lines

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 2]
```diff
@@ -2233,19 +2234,19 @@ subroutine Rtmrun(rstwr,nlend,rdate)
...
-       if (inundflag) then
-          call t_startf('mosartr_inund_sim')
-          call MOSARTinund_simulate ( )
-          call t_stopf('mosartr_inund_sim')
-       else
+       !if (inundflag .and. wrmflag .eq. 0) then !use Luo's scheme when inundation is on and WM is off (keep it for now - tz)
+       !   call t_startf('mosartr_inund_sim')
+       !   call MOSARTinund_simulate ( )
+       !   call t_stopf('mosartr_inund_sim')
+       !else ! other cases
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: delete commented out lines and re-indent code

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 3]
```diff
@@ -2233,19 +2234,19 @@ subroutine Rtmrun(rstwr,nlend,rdate)
...
           call t_startf('mosartr_euler')
           ! debug 
 #ifdef DEBUG
           write(iulog,*) 'clm-mosart subT: (call Euler) ns=', ns
 #endif
           call Euler()
           call t_stopf('mosartr_euler')
-       endif
+       !endif
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: delete commented out line

File: components/mosart/src/wrm/WRM_subw_IO_mod.F90 [hunk 4]
```diff
@@ -753,17 +753,16 @@ subroutine WRM_init
...
+          ! if (masterproc) write(iulog,FORMR) trim(subname),'prerelease Jul',minval(StorWater%pre_release(:,7)),maxval(StorWater%pre_release(:,7))
+          ! if (masterproc) write(iulog,FORMR) trim(subname),'prerelease Oct', minval(StorWater%pre_release(:,10)),maxval(StorWater%pre_release(:,10))
+          ! if (masterproc) write(iulog,FORMR) trim(subname),'Coulee',StorWater%pre_release(80,1),StorWater%pre_release(80,3)
+          ! if (masterproc) write(iulog,FORMR) trim(subname),'Coulee',StorWater%pre_release(80,5),StorWater%pre_release(80,9)
+          ! if (masterproc) write(iulog,FORMR) trim(subname),'Coulee Flow',WRMUnit%MeanMthFlow(80,13), WRMUnit%MeanMthFlow(80,8)
+          ! if (masterproc) write(iulog,FORMR) trim(subname),'Coulee Demand',WRMUnit%MeanMthFlow(80,1),WRMUnit%MeanMthFlow(80,3)
+          ! if (masterproc) write(iulog,FORMR) trim(subname),'Coulee Demand',WRMUnit%MeanMthFlow(80,5),WRMUnit%MeanMthFlow(80,7)
+          ! if (masterproc) write(iulog,FORMR) trim(subname),'Coulee Means Flow Demand',WRMUnit%MeanMthDemand(80,13),WRMUnit%MeanMthFlow(80,13)
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_subw_IO_mod.F90]: delete commented out lines

File: components/mosart/src/wrm/WRM_subw_IO_mod.F90 [hunk 5]
```diff
@@ -803,10 +802,9 @@ subroutine WRM_readDemand
      if (ctlSubwWRM%ExtractionFlag > 0) then
 
         call get_curr_date(yr, mon, day, tod)
-        write(iulog,'(2a,4i6)') subname,'at ',yr,mon,day,tod
-    
-        write(strYear,'(I4.4)') yr
-        write(strMonth,'(I2.2)') mon
+      !  write(iulog,'(2a,4i6)') subname,'at ',yr,mon,day,tod    
+      !  write(strYear,'(I4.4)') yr
+      !  write(strMonth,'(I2.2)') mon
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_subw_IO_mod.F90]: delete commented out lines

File: components/mosart/src/wrm/WRM_subw_IO_mod.F90 [hunk 6]
```diff
@@ -844,18 +842,18 @@ subroutine WRM_computeRelease()
...
      call RegulationRelease()
-     write(iulog,*) 'Start Coulee ',mon,day,tod,WRMUnit%MeanMthFlow(80,13)
-     write(iulog,*) 'start Op mon, storage ', WRMUnit%MthStOp(80),WRMUnit%StorMthStOp(80)
-     write(iulog,*)  'storage, release pre targets ',StorWater%storage(80), StorWater%release(80)
+    ! write(iulog,*) 'Start Coulee ',mon,day,tod,WRMUnit%MeanMthFlow(80,13)
+    ! write(iulog,*) 'start Op mon, storage ', WRMUnit%MthStOp(80),WRMUnit%StorMthStOp(80)
+    ! write(iulog,*)  'storage, release pre targets ',StorWater%storage(80), StorWater%release(80)
      call WRM_storage_targets()
-     write(iulog,*) 'Coulee targets ',StorWater%release(80)
+    ! write(iulog,*) 'Coulee targets ',StorWater%release(80)
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_subw_IO_mod.F90]: delete commented out lines

File: components/clm/src/biogeophys/CanopyHydrologyMod.F90 [hunk 7]
```diff
@@ -440,69 +374,50 @@ subroutine CanopyHydrology(bounds, &
...
+               qflx_real_irrig(p) = 0._r8 
+               qflx_surf_irrig(p) = 0._r8
+               qflx_grnd_irrig(p) = 0._r8
+               qflx_over_supply(p) = 0._r8
+
+             if (qflx_irrig(p) > 0._r8 .or. qflx_supply(p) > 0._r8) then	!this pft needs water or have supply             
+               if  (irrigated(veg_pp%itype(p)) == 1._r8) then ! this pft is irrigated
+                 !if (irrigated_ppg(g)<1) then
+                 !   write (iulog,*) 'DEBUG irrigated_ppg(g)', irrigated_ppg(g)
+                 !endif
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyHydrologyMod.F90]: delete commented out lines

File: components/clm/src/biogeophys/CanopyHydrologyMod.F90 [hunk 8]
```diff
@@ -258,77 +253,18 @@ subroutine CanopyHydrology(bounds, &
...
-            qflx_irrig_grid(gg) = 0
-         end do
-        
-        ! loop the pft and assign irrig_rate to qflx_irrig based on n_irrig_steps_left(p), 
-        ! which will be zero if current pft doesn't need irrigation or the current time 
-        ! is out of irrigation schedule
-         do f = 1, num_nolakep
+
+       do gg = bounds%begg,bounds%endg
+          irrigated_ppg(gg) = 0
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyHydrologyMod.F90]: change `0` to `0._r8`

File: components/mosart/src/riverroute/RtmMod.F90 (hunk 2 above)
Comment (hydrotian) [reply components/mosart/src/riverroute/RtmMod.F90]: This is the old scheme. I intend to keep it for now as we may do some comparisons in the future.

**Comment by dqwu** [comment]
@bishtgautam 
This PR has been merged to next branch but there is a typo in the merge commit message (the PR number should be 3441, not 3341)

Also, this PR has caused a test to fail on melvin:
https://my.cdash.org/viewTest.php?onlyfailed&buildid=1785671

I can reproduce this failure on anlworkstation, and the error message is:
```
At line 191 of file components/clm/src/cpl/lnd_import_export.F90
Fortran runtime error: Index '0' of dimension 1 of array 'x2l' below lower bound of 1
```

**Comment by bishtgautam** [comment]
@dqwu Thanks for catching the typo and reproducing the failure on anlworkstation.

@hydrotian, Can you fix the error as pointed out by @dqwu?

**Comment by hydrotian** [comment]
@bishtgautam I made a change but not sure if this can fix the problem - I don't have access to GNU machine. Maybe @dqwu can test it again? Thanks.

**Comment by dqwu** [comment]
@hydrotian 
SMS_D_Ln5.ne4_ne4.FC5AV1C-L.anlworkstation_gnu passed with your updated PR.
I can be re-merged to next for testing on other machines.

**Comment by bishtgautam** [comment]
@dqwu, Thanks for testing out the new changes.

The branch has been remerged to next.

**Comment by rljacob** [comment]
@hydrotian you should get an account on anvil which has gnu.

**Comment by hydrotian** [comment]
Thanks to @dqwu and @bishtgautam . @rljacob , who should I talk to to get anvil account?

**Comment by rljacob** [comment]
@hydrotian that info is here: https://acme-climate.atlassian.net/wiki/spaces/ED/pages/98992379/Anvil+-+ACME+s+dedicated+nodes+hosted+on+Blues#Anvil-ACME%27sdedicatednodeshostedonBlues-GettinganAccount

**Comment by hydrotian** [comment]
Thanks @rljacob

**Pull request merged**